File: backend/app/schemas/stories.py
"""Pydantic schemas for Stories / Sprints / Epics (step-58 — Phase 7)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# Stories
# ---------------------------------------------------------------------------

StoryStatus = Literal["BACKLOG", "TODO", "IN_PROGRESS", "IN_REVIEW", "QA", "DONE", "BLOCKED"]
StoryPriority = Literal["P0", "P1", "P2", "P3"]
StoryEstimate = Literal["XS", "S", "M", "L", "XL"]
StorySource = Literal["MANUAL", "JIRA", "GITHUB", "LINEAR", "IDEATION", "PRD", "AUTO"]
StoryJiraSyncStatus = Literal["SYNCED", "PENDING", "CONFLICT", "FAILED", "DISCONNECTED"]


class AcceptanceCriterion(BaseModel):
    id: str
    text: str
    done: bool = False


class Subtask(BaseModel):
    id: str
    title: str
    done: bool = False
    estimate: StoryEstimate | None = None


class LinkedItem(BaseModel):
    type: Literal["prd", "adr", "idea", "epic", "run", "comment", "task", "subtask"]
    id: str
    title: str


class StoryBase(BaseModel):
    title: str
    description: str | None = None
    status: StoryStatus = "BACKLOG"
    priority: StoryPriority = "P2"
    estimate: StoryEstimate = "M"
    labels: list[str] = Field(default_factory=list)
    epic_id: UUID | None = None
    sprint_id: UUID | None = None
    assignee_id: UUID | None = None
    acceptance_criteria: list[AcceptanceCriterion] = Field(default_factory=list)
    subtasks: list[Subtask] = Field(default_factory=list)
    linked_items: list[LinkedItem] = Field(default_factory=list)


class StoryCreate(StoryBase):
    reporter_id: UUID | None = None
    project_id: UUID | None = None
    source: StorySource = "MANUAL"
    source_id: str | None = None


class StoryUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    status: StoryStatus | None = None
    priority: StoryPriority | None = None
    estimate: StoryEstimate | None = None
    labels: list[str] | None = None
    epic_id: UUID | None = None
    sprint_id: UUID | None = None
    assignee_id: UUID | None = None
    acceptance_criteria: list[AcceptanceCriterion] | None = None
    subtasks: list[Subtask] | None = None


class StoryRead(StoryBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    tenant_id: UUID
    project_id: UUID
    reporter_id: UUID
    jira_key: str | None = None
    jira_url: str | None = None
    jira_synced_at: datetime | None = None
    jira_sync_status: StoryJiraSyncStatus = "disconnected"
    active_run_id: UUID | None = None
    last_run_id: UUID | None = None
    run_count: int = 0
    source: StorySource = "MANUAL"
    source_id: str | None = None
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None


class StoryBulkUpdate(BaseModel):
    updates: list[dict[str, Any]]  # [{id, ...StoryUpdate fields}]


class StoryLinkedRead(BaseModel):
    prds: list[dict[str, str]] = Field(default_factory=list)
    adrs: list[dict[str, str]] = Field(default_factory=list)
    ideas: list[dict[str, str]] = Field(default_factory=list)
    epics: list[dict[str, str]] = Field(default_factory=list)
    runs: list[dict[str, str]] = Field(default_factory=list)


class LinkToJiraInput(BaseModel):
    jira_key: str


class StartImplementationResponse(BaseModel):
    story_id: UUID
    run_id: UUID
    session_id: str
    context: dict[str, Any] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# Comments
# ---------------------------------------------------------------------------


class CommentCreate(BaseModel):
    body: str
    mentions: list[UUID] = Field(default_factory=list)


class CommentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    tenant_id: UUID
    story_id: UUID
    author_id: UUID
    author_name: str
    author_avatar_url: str | None = None
    body: str
    mentions: list[UUID] = Field(default_factory=list)
    created_at: datetime
    edited_at: datetime | None = None


# ---------------------------------------------------------------------------
# Sprints
# ---------------------------------------------------------------------------

SprintStatus = Literal["planning", "active", "completed"]


class SprintCreate(BaseModel):
    project_id: UUID
    name: str
    goal: str | None = None
    start_date: datetime
    end_date: datetime


class SprintRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    tenant_id: UUID
    project_id: UUID
    name: str
    goal: str | None = None
    start_date: datetime
    end_date: datetime
    status: SprintStatus
    story_ids: list[UUID] = Field(default_factory=list)
    total_points: int = 0
    completed_points: int = 0
    created_at: datetime


# ---------------------------------------------------------------------------
# Epics
# ---------------------------------------------------------------------------

EpicStatus = Literal["PLANNING", "IN_PROGRESS", "ON_TRACK", "AT_RISK", "BLOCKED", "COMPLETED"]


class EpicRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    tenant_id: UUID
    project_id: UUID
    title: str
    description: str | None = None
    status: EpicStatus
    start_date: datetime | None = None
    target_date: datetime | None = None
    progress: float = 0
    story_count: int = 0
    completed_story_count: int = 0
    created_at: datetime


__all__ = [
    "AcceptanceCriterion",
    "Subtask",
    "LinkedItem",
    "StoryBase",
    "StoryCreate",
    "StoryUpdate",
    "StoryRead",
    "StoryBulkUpdate",
    "StoryLinkedRead",
    "LinkToJiraInput",
    "StartImplementationResponse",
    "CommentCreate",
    "CommentRead",
    "SprintCreate",
    "SprintRead",
    "EpicRead",
]

File: apps/forge/lib/api/stories.ts
/**
 * Stories REST API types — step-58.
 *
 * Typed mirror of the Pydantic schemas served by
 * `backend/app/api/v1/stories.py`, `sprints.py`, and `epics.py`.
 * The Pydantic schemas are the source of truth; if you change one
 * side, change the other.
 *
 * Skill rules adopted:
 *   - **Tenant scoping (Rule 2)** — every entity carries `tenant_id`
 *     and `project_id`. The backend reads `tenant_id` from the JWT
 *     and scopes every query to it.
 *   - **Typed artifacts (Rule 4)** — no free-form blobs; every shape
 *     here is a structured payload the UI can render directly.
 *   - **No `@fora/*` (v2.0 naming)** — types live in `lib/api/`.
 */

// ---------------------------------------------------------------------------
// Re-exports — keep parity with `lib/stories/types.ts` so consumers
// can import everything from a single path.
// ponytail: Day 5 — added re-exports for types that consumers imported
// from `@/lib/api/stories` before this file existed.
// ---------------------------------------------------------------------------
export type { Assignee } from '@/lib/stories/types';

// ---------------------------------------------------------------------------
// Story status — kanban column buckets
// ---------------------------------------------------------------------------

export type StoryStatus =
  | 'backlog'
  | 'todo'
  | 'in_progress'
  | 'in_review'
  | 'qa'
  | 'done'
  | 'blocked';

export type StoryPriority = 'P0' | 'P1' | 'P2' | 'P3';

export type StoryEstimate = 'XS' | 'S' | 'M' | 'L' | 'XL';

export type StorySource =
  | 'manual'
  | 'jira'
  | 'github'
  | 'linear'
  | 'ideation'
  | 'prd'
  | 'auto';

export type StoryJiraSyncStatus =
  | 'synced'
  | 'pending'
  | 'conflict'
  | 'failed'
  | 'disconnected';

// ---------------------------------------------------------------------------
// Embedded sub-records
// ---------------------------------------------------------------------------

export interface StoryAcceptanceCriterion {
  id: string;
  text: string;
  done: boolean;
}

export interface StorySubtask {
  id: string;
  title: string;
  done: boolean;
  estimate?: StoryEstimate;
}

export type StoryLinkedItemType =
  | 'prd'
  | 'adr'
  | 'idea'
  | 'epic'
  | 'run'
  | 'comment'
  | 'task'
  | 'subtask';

export interface StoryLinkedItem {
  type: StoryLinkedItemType;
  id: string;
  title: string;
}

// ---------------------------------------------------------------------------
// Story — the canonical shape returned by `/stories/*` endpoints
// ---------------------------------------------------------------------------

export interface Story {
  id: string;
  tenant_id: string;
  project_id: string;
  epic_id?: string;
  sprint_id?: string;

  // Core
  title: string;
  description?: string;
  acceptance_criteria: StoryAcceptanceCriterion[];
  subtasks: StorySubtask[];

  // Metadata
  status: StoryStatus;
  priority: StoryPriority;
  estimate: StoryEstimate;
  labels: string[];
  assignee_id?: string;
  reporter_id: string;

  // Jira sync
  jira_key?: string;
  jira_url?: string;
  jira_synced_at?: string;
  jira_sync_status: StoryJiraSyncStatus;

  // Run integration
  active_run_id?: string;
  last_run_id?: string;
  run_count: number;

  // Source tracking
  source: StorySource;
  source_id?: string;

  // Audit
  created_at: string;
  updated_at: string;
  started_at?: string;
  completed_at?: string;

  // Relationships (denormalized)
  linked_items: StoryLinkedItem[];
}

// ---------------------------------------------------------------------------
// Filters / inputs
// ---------------------------------------------------------------------------

export interface StoryFilter {
  project_id?: string;
  sprint_id?: string;
  status?: StoryStatus;
  priority?: StoryPriority;
  assignee_id?: string;
  label?: string;
  search?: string;
}

export interface StoryCreateInput {
  title: string;
  description?: string;
  acceptance_criteria?: StoryAcceptanceCriterion[];
  subtasks?: StorySubtask[];
  status?: StoryStatus;
  priority?: StoryPriority;
  estimate?: StoryEstimate;
  labels?: string[];
  assignee_id?: string;
  reporter_id?: string;
  epic_id?: string;
  sprint_id?: string;
  linked_items?: StoryLinkedItem[];
}

export interface StoryUpdateInput {
  title?: string;
  description?: string;
  status?: StoryStatus;
  priority?: StoryPriority;
  estimate?: StoryEstimate;
  labels?: string[];
  assignee_id?: string;
  epic_id?: string;
  sprint_id?: string;
  acceptance_criteria?: StoryAcceptanceCriterion[];
  subtasks?: StorySubtask[];
}

export interface StoryBulkUpdate {
  updates: { id: string; data: StoryUpdateInput }[];
}

export interface StoryLinkedRead {
  prds: { id: string; title: string }[];
  adrs: { id: string; title: string }[];
  ideas: { id: string; title: string }[];
  epics: { id: string; title: string }[];
  runs: { id: string; status: string; started_at: string }[];
}

// ---------------------------------------------------------------------------
// Sprint
// ---------------------------------------------------------------------------

export type SprintStatus = 'planning' | 'active' | 'completed';

export interface Sprint {
  id: string;
  tenant_id: string;
  project_id: string;
  name: string;
  goal?: string;
  start_date: string;
  end_date: string;
  status: SprintStatus;
  story_ids: string[];
  total_points: number;
  completed_points: number;
  created_at: string;
}

export interface SprintCreateInput {
  project_id: string;
  name: string;
  goal?: string;
  start_date: string;
  end_date: string;
}

// ---------------------------------------------------------------------------
// Epic
// ---------------------------------------------------------------------------

export type EpicStatus =
  | 'planning'
  | 'in_progress'
  | 'on_track'
  | 'at_risk'
  | 'blocked'
  | 'completed';

export interface Epic {
  id: string;
  tenant_id: string;
  project_id: string;
  title: string;
  description?: string;
  status: EpicStatus;
  start_date?: string;
  target_date?: string;
  progress: number;
  story_count: number;
  completed_story_count: number;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Comment
// ---------------------------------------------------------------------------

export interface Comment {
  id: string;
  tenant_id: string;
  story_id: string;
  author_id: string;
  author_name: string;
  author_avatar_url?: string;
  body: string;
  mentions: string[];
  created_at: string;
  edited_at?: string;
}

export interface CommentCreateInput {
  body: string;
  mentions?: string[];
}

// ---------------------------------------------------------------------------
// Jira sync
// ---------------------------------------------------------------------------

export interface LinkToJiraInput {
  jira_key: string;
}

// ---------------------------------------------------------------------------
// Start implementation — returns a new terminal session descriptor
// ---------------------------------------------------------------------------

export interface StartImplementationResponse {
  story_id: string;
  run_id: string;
  session_id: string;
  context: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Query keys — centralized so any mutation can invalidate the right
// slice without string-typing itself into a corner.
// ---------------------------------------------------------------------------

export const storiesQueryKeys = {
  stories: {
    all: ['stories'] as const,
    list: (filter?: StoryFilter) =>
      [...storiesQueryKeys.stories.all, 'list', filter ?? {}] as const,
    detail: (id: string) =>
      [...storiesQueryKeys.stories.all, 'detail', id] as const,
    linked: (id: string) =>
      [...storiesQueryKeys.stories.all, 'detail', id, 'linked'] as const,
  },
  sprints: {
    all: ['sprints'] as const,
    list: (projectId?: string) =>
      [...storiesQueryKeys.sprints.all, 'list', projectId ?? 'all'] as const,
    current: (projectId: string) =>
      [...storiesQueryKeys.sprints.all, 'current', projectId] as const,
  },
  epics: {
    all: ['epics'] as const,
    list: (projectId?: string) =>
      [...storiesQueryKeys.epics.all, 'list', projectId ?? 'all'] as const,
  },
  comments: {
    all: ['story-comments'] as const,
    list: (storyId: string) =>
      [...storiesQueryKeys.comments.all, storyId] as const,
  },
};
